src/core/algorithms/nar/des/encoded_nar.cpp

```diff
@@ -0,0 +1,102 @@
+#include "encoded_nar.h"
+
+#include "algorithms/nar/value_range.h"
+#include "model/types/types.h"
+
+namespace algos::des {
+
+using model::NAR;
+
+size_t EncodedNAR::VectorSize() {
+    return encoded_value_ranges_.size() * EncodedValueRange().kFieldCount + 1;
+}
+
+size_t EncodedNAR::FeatureCount() {
+    return encoded_value_ranges_.size();
+}
+
+// TODO: remove code duplication here
+double& EncodedNAR::operator[](size_t index) {
+    qualities_consistent_ = false;
+    if (index == 0) {
+        return implication_sign_pos_;
+    } else {
+        index--;
+        size_t feature = index / EncodedValueRange().kFieldCount;
+        size_t feature_field = index % EncodedValueRange().kFieldCount;
+        return encoded_value_ranges_[feature][feature_field];
+    }
+}
+
+double const& EncodedNAR::operator[](size_t index) const {
+    if (index == 0) {
+        return implication_sign_pos_;
+    } else {
+        index--;
+        size_t feature = index / EncodedValueRange().kFieldCount;
+        size_t feature_field = index % EncodedValueRange().kFieldCount;
+        return encoded_value_ranges_[feature][feature_field];
+    }
+}
+
+NAR EncodedNAR::SetQualities(FeatureDomains domains, TypedRelation const* typed_relation) {
+    NAR this_decoded = Decode(domains);
+    this_decoded.SetQualities(typed_relation);
+    qualities_ = this_decoded.GetQualities();
+    qualities_consistent_ = true;
+    return this_decoded;
+}
+
+model::NARQualities const& EncodedNAR::GetQualities() const {
+    if (!qualities_consistent_) {
+        throw std::logic_error("Getting uninitialized qualities from NAR.");
+    }
+    return qualities_;
+}
+
+NAR EncodedNAR::Decode(FeatureDomains domains) const {
+    auto resulting_nar = model::NAR();
+
+    std::vector<size_t> feature_order(encoded_value_ranges_.size());
+    std::iota(std::begin(feature_order), std::end(feature_order), 0);
+    auto compare_by_permutation = [e = encoded_value_ranges_](size_t& a, size_t& b) -> bool {
+        return e[a].permutation > e[b].permutation;
+    };
+    std::sort(feature_order.begin(), feature_order.end(), compare_by_permutation);
+
+    uint implication_sign_after = implication_sign_pos_ * (encoded_value_ranges_.size() - 1);
+    uint handling_feat_num = -1;
+    for (size_t feature_index : feature_order) {
+        handling_feat_num++;
+        EncodedValueRange encoded_feature = encoded_value_ranges_[feature_index];
+        if (encoded_feature.threshold < RNG().Next()) {
+            continue;
+        }
+        auto domain = domains[feature_index];
+        auto decoded = encoded_value_ranges_[feature_index].Decode(domain);
```

Comment:
again, you copying the shared pointer when using `auto domain = domains[feature_index]`. Use const reference, no need to increase ref count of shared ptr for no reason

Reply:
again, no need to create an additional copy of `decoded`.
I'm aware that  `InsertInCons` wants a shared pointer, not a const ref, but when passing const ref to it it will create a copy automatically, so no problems here.

Reply:
If we want to avoid all copies why don't we use `std::move` everywhere and avoid copies altogether:
```c++
auto domain = std::move(domains[feature_index]);
auto decoded = std::move(encoded_value_ranges_[feature_index].Decode(std::move(domain)));
if (handling_feat_num > implication_sign_after) {
            resulting_nar.InsertInCons(feature_index, std::move(decoded);)
        } else {
            resulting_nar.InsertInAnte(feature_index, std::move(decoded);)
        }
. . .
```
my types only use basic C++ types so their generated move constructors should be valid? To my understanding we don't because a) it makes code less readable and b) compiler should be able to figure out where to move most basic objects on its own.